basics/uri.adoc

```diff
@@ -25,404 +26,522 @@ specific language governing permissions and limitations
-* *MUST* return `UStatus::Failed` if the `UUri.entity.name` is blank
+[mermaid]
+ifdef::env-github[[source,mermaid]]
+----
+classDiagram
+
+class UUri {
+  authority_name : String
+  ue_id : UInt32
+  ue_version_major : UInt8
+  resource_id : UInt16
+}
+----
 
-| `validateRpcMethod(UUri): UStatus`
-a| Validate that the `UUri` is a valid RPC method UUri
+[.specitem,oft-sid="dsn~data-model-naming~1",oft-needs="impl"]
+--
+Each link:../languages.adoc[uProtocol Language Library] *MUST* implement the data model using the type and property names as defined in the following sections.
+--
 
-* *MUST*  return `UStatus::Failed` if `validate(UUri)` fails or `isRpcMethod()` is false
+[.specitem,oft-sid="req~data-model-proto~1",oft-needs="impl,utest"]
+--
+Each link:../languages.adoc[uProtocol Language Library] *MUST* support writing and reading of an instance of the data model to/from a protobuf as defined in link:../up-core-api/uprotocol/uri.proto[uri.proto].
+--
 
-| `validateRpcResponse(UUri): UStatus`
-a| Validate that the `UUri` is a valid RPC Response UUri 
+The remainder of this section defines the requirements for the data model by means of _invariants_ which must hold true at all times. Additionally, there are some _predicates_ that can be used by uEntities to determine if a UUri contains the information they are interested in. Both the invariants and the predicates are (formally) defined by means of link:https://www.omg.org/spec/OCL/2.4[Object Constraint Language] (OCL) expressions on the UUri data model. The link:https://modeling-languages.com/ocl-tutorial/[OCL Tutorial] provides an overview of the basic concepts used in the definitions.
 
-* *MUST*  return `UStatus::Failed` if  `validate(UUri)` or `validateRpcMethod(UUri)` fails or `isRpcResponse()` is false
 
-| `isLocal(UUri): boolean`
-| Return true if: `!UUri.hasAuthority()`
+=== Authority
 
-| `isLongForm(UUri): boolean`
-a| Returns true if the Uri part contains names so that it can be serialized to <<Long URIs>>.
+An authority represents the deployment location of a specific software entity. The location is represented by means of a logical identifier like a domain name (e.g. `mcu1.example.com`) or a Vehicle Identification Number (VIN).
 
-| `isEmpty(UUri): boolean`
-| Returns true if the object contains no information
 
-| `isResolved(UUri): boolean`
-| Returns true if the object contains both names and numbers (ids) such that it can be serialize to both <<Long URIs>> and <<Micro URIs>>
+[.specitem,oft-sid="dsn~authority-name-length~1",oft-needs="impl,utest"]
+--
+A UUri's `authority_name` *MUST NOT* exceed 128 characters in length.
 
-| `isMicroForm(UUri): boolean`
-| Returns true if the Uri part contains the required ids to serialized to Micro format and the fields of the Uri can fit within the specified number of bits and bytes. Both sets of details can be obtained under <<Micro URIs>>
+[source,ocl]
+----
+inv: authority_name.length() <= 128
```

Comment:
Where did this requirement come from?

Reply:
We had been discussing that we wanted to limit the length of the authority name to some degree. I just picked 128 as a first shot. Do you still think that we should limit the length?